### PR TITLE
Feature/sg 13 coco dataset yolox

### PR DESCRIPTION
**Pull request by shaydeci**
COCO detection dataset interface for YOLOX.

Relies on https://github.com/Deci-AI/super-gradients/pull/208 to be merged first.

### CODE REVIEW

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        width = im_ann["width"]
+        height = im_ann["height"]
+        anno_ids = self.coco.getAnnIds(imgIds=[int(id_)], iscrowd=False)
+        annotations = self.coco.loadAnns(anno_ids)
+        objs = []
+        for obj in annotations:
+            x1 = np.max((0, obj["bbox"][0]))
+            y1 = np.max((0, obj["bbox"][1]))
+            x2 = np.min((width, x1 + np.max((0, obj["bbox"][2]))))
+            y2 = np.min((height, y1 + np.max((0, obj["bbox"][3]))))
```
Comment (ofrimasad) [inline src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:83]: what is the logic here? we don't trust the annotations file to be accurate? x2-x1 can be different than width?

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+                "Caching images for the first time. This might take about 20 minutes for COCO"
+            )
+            self.imgs = np.memmap(
+                cache_file,
+                shape=(len(self.ids), max_h, max_w, 3),
+                dtype=np.uint8,
+                mode="w+",
+            )
+            from tqdm import tqdm
+            from multiprocessing.pool import ThreadPool
```
Comment (ofrimasad) [inline src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:145]: why is this import here and not at the top?

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+            NUM_THREADs = min(8, os.cpu_count())
+            loaded_images = ThreadPool(NUM_THREADs).imap(
```
Comment (ofrimasad) [inline src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:148]: i think using a Threadpool inside a distributed dataloader may cause issues.

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        )
+
+        return res, res_seg, img_info, resized_info, os.path.join(self.data_dir, self.name, file_name)
+
+    def _cache_images(self):
+        logger.warning(
+            "\n********************************************************************************\n"
+            "You are using cached images in RAM to accelerate training.\n"
+            "This requires large system RAM.\n"
+            "Make sure you have 200G+ RAM and 136G available disk space for training COCO.\n"
```
Comment (ofrimasad) [inline src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:128]: an important thing to note. if you are using DDP, this will be 8 times more memory!

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        """
+        id_ = self.ids[index]
+        res, res_seg, img_info, resized_info, _ = self.annotations[index]
+        if self.imgs is not None:
+            pad_img = self.imgs[index]
+            img = pad_img[: resized_info[0], : resized_info[1], :].copy()
+        else:
+            img = self.load_resized_img(index)
+
+        sample = {"image": img, "target": res.copy(), "target_seg": res_seg, "info": img_info, "id": np.array([id_])}
```
Comment (ofrimasad) [inline src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:209]: will the target include the "crowd" parameter?

File: src/super_gradients/training/datasets/samplers/infinite_sampler.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
```
Comment (ofrimasad) [inline src/super_gradients/training/datasets/samplers/infinite_sampler.py:1]: please remove

File: src/super_gradients/training/datasets/samplers/infinite_sampler.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+# -*- coding:utf-8 -*-
+# Copyright (c) Megvii, Inc. and its affiliates.
```
Comment (ofrimasad) [inline src/super_gradients/training/datasets/samplers/infinite_sampler.py:3]: what? what license is this?

File: src/super_gradients/training/utils/detection_utils.py [hunk 8]
```diff
@@ -1280,3 +1280,15 @@ def __call__(self, data):
             targets_merged.append(torch.cat((batch_column, targets_im), 1))
         targets = torch.cat(targets_merged, 0)
         return ims, targets
+
+def get_yolox_datadir():
```
Comment (ofrimasad) [inline src/super_gradients/training/utils/detection_utils.py:1284]: please remove

File: src/super_gradients/training/sg_model/sg_model.py [hunk 9]
```diff
@@ -911,7 +917,7 @@ def train(self, training_params: dict = dict()):  # noqa: C901
 
                 # IN DDP- SET_EPOCH WILL CAUSE EVERY PROCESS TO BE EXPOSED TO THE ENTIRE DATASET BY SHUFFLING WITH A
                 # DIFFERENT SEED EACH EPOCH START
-                if self.multi_gpu == MultiGPUMode.DISTRIBUTED_DATA_PARALLEL:
+                if self.multi_gpu == MultiGPUMode.DISTRIBUTED_DATA_PARALLEL and hasattr(self.train_loader, "sampler") and hasattr(self.train_loader.sampler, "set_epoch"):
```
Comment (ofrimasad) [inline src/super_gradients/training/sg_model/sg_model.py:920]: i don't like this logic. this sub-sub-parameter is overriding the multi GPU mode?

File: src/super_gradients/training/utils/distributed_training_utils.py [hunk 10]
```diff
@@ -114,3 +116,40 @@ def compute_precise_bn_stats(model: nn.Module, loader: torch.utils.data.DataLoad
         bn.running_mean = running_means[i]
         bn.running_var = running_vars[i]
         bn.momentum = momentums[i]
+
+
+def get_local_rank():
```
Comment (ofrimasad) [inline src/super_gradients/training/utils/distributed_training_utils.py:121]: why did you add this on this PR? i don't see it used anywhere

File: src/super_gradients/training/utils/distributed_training_utils.py [hunk 11]
```diff
@@ -114,3 +116,40 @@ def compute_precise_bn_stats(model: nn.Module, loader: torch.utils.data.DataLoad
...
+    """
+    if not dist.is_available():
+        return 1
+    if not dist.is_initialized():
+        return 1
+    return dist.get_world_size()
+
+
+@contextmanager
+def wait_for_the_master(local_rank: int):
```
Comment (ofrimasad) [inline src/super_gradients/training/utils/distributed_training_utils.py:142]: why did you add this on this PR? i don't see it used anywhere

File: src/super_gradients/training/utils/distributed_training_utils.py [hunk 12]
```diff
@@ -114,3 +116,40 @@ def compute_precise_bn_stats(model: nn.Module, loader: torch.utils.data.DataLoad
...
+
+def get_local_rank():
+    """
+    Returns the local rank if running in DDP, and 0 otherwise
+    :return: local rank
+    """
+    return dist.get_rank() if dist.is_initialized() else 0
+
+
+def get_world_size() -> int:
```
Comment (ofrimasad) [inline src/super_gradients/training/utils/distributed_training_utils.py:129]: why did you add this on this PR? i don't see it used anywhere

**Review by ofrimasad — COMMENTED** [review]
it looks like some of the code changes here are not really code changes but actually, just formatting changes generated by your misconfigured pycharm. please try to avoid those (especially in such large PRs).

other comments inline

File: src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py [hunk 13]
```diff
@@ -879,3 +892,97 @@ def __init__(self, dataset_params=None, cache_labels: bool = False, cache_images
         )
 
         self.classes = self.trainset.classes
+
+
+class CocoDetectionDatasetInterfaceYolox(DatasetInterface):
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py]: done

File: src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py [hunk 14]
```diff
@@ -879,3 +892,97 @@ def __init__(self, dataset_params=None, cache_labels: bool = False, cache_images
...
+    def __init__(self, dataset_params={}):
+        super(CocoDetectionDatasetInterfaceYolox, self).__init__()
+        default_coco_dataset_params = {"mixup_prob": 1.0,
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py]: done in https://github.com/Deci-AI/super-gradients/pull/223/commits/65ea51f037016daae86f248c875ee72ba7f1f2c2 , also added the yaml for transparency.

File: src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py [hunk 15]
```diff
@@ -879,3 +892,97 @@ def __init__(self, dataset_params=None, cache_labels: bool = False, cache_images
...
+                            DetectionMixup(input_dim=train_input_dim,
+                                           mixup_scale=self.dataset_params.mixup_scale,
+                                           prob=self.dataset_params.mixup_prob),
+                            DetectionHSV(prob=self.dataset_params.hsv_prob),
+                            DetectionHorizontalFlip(prob=self.dataset_params.flip_prob),
+                            DetectionPaddedRescale(input_dim=train_input_dim, max_targets=120),
+                            DetectionTargetsFormatTransform()
+                            ]
+
+        self.trainset = COCODetectionDatasetYolox(data_dir="/data/coco",
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py]: done in https://github.com/Deci-AI/super-gradients/commit/65ea51f037016daae86f248c875ee72ba7f1f2c2 , also added the yaml for transparency.

File: src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py [hunk 16]
```diff
@@ -879,3 +892,97 @@ def __init__(self, dataset_params=None, cache_labels: bool = False, cache_images
...
+                                                  json_file="instances_train2017.json",
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py]: done in https://github.com/Deci-AI/super-gradients/commit/65ea51f037016daae86f248c875ee72ba7f1f2c2 , also added the yaml for transparency.

File: src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py [hunk 17]
```diff
@@ -879,3 +892,97 @@ def __init__(self, dataset_params=None, cache_labels: bool = False, cache_images
...
+                                                  img_size=(self.dataset_params.train_image_size,
+                                                            self.dataset_params.train_image_size),
+                                                  cache=self.dataset_params.cache_images,
+                                                  transforms=train_transforms
+                                                  )
+
+        val_input_dim = (self.dataset_params.val_image_size, self.dataset_params.val_image_size)
+        self.valset = COCODetectionDatasetYolox(
+            data_dir='/data/coco',
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/dataset_interfaces/dataset_interface.py]: done in https://github.com/Deci-AI/super-gradients/commit/65ea51f037016daae86f248c875ee72ba7f1f2c2 , also added the yaml for transparency.

File: src/super_gradients/training/datasets/datasets_utils.py [hunk 18]
```diff
@@ -529,3 +531,10 @@ def get_color_augmentation(rand_augment_config_string: str, color_jitter: tuple,
     else:  # RandAugment includes colorjitter like augmentations, both cannot be applied together.
         color_augmentation = transforms.ColorJitter(*color_jitter)
     return color_augmentation
+
+
+def worker_init_reset_seed(worker_id):
+    seed = uuid.uuid4().int % 2 ** 32
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/datasets_utils.py:537]: Reason added in method docs.

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 19]
```diff
@@ -0,0 +1,287 @@
+import os
+
+from pycocotools.coco import COCO
+from super_gradients.common.abstractions.abstract_logger import get_logger
+from torch.utils.data import Dataset
+import random
+
+import cv2
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:8]: I didn't encounter such thing when training yolox...
We can add cv2.setNumThreads(0), but it would probably effect performance.

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+import numpy as np
+
+from super_gradients.training.utils.detection_utils import get_yolox_datadir
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py]: removed

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+
+logger = get_logger(__name__)
+
+
+class COCODetectionDatasetYolox(Dataset):
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py]: done

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+    """Detection dataset COCO implementation"""
+
+    def __init__(
+            self, img_size: tuple = (640, 640),
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py]: Removed

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 23]
```diff
@@ -0,0 +1,287 @@
...
+        :param tight_box_rotation: bool, whether to use of segmentation maps convex hull
+         as target_seg (see load_sample docs).
+        :param transforms: list of transforms to apply sequentially on sample in __getitem__
+        """
+        super().__init__()
+        self.imgs = None
+        self.data_dir = data_dir
+        self.input_dim = img_size
+        if data_dir is None:
+            data_dir = os.path.join(get_yolox_datadir(), "COCO")
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py]: Removed

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+        self.data_dir = data_dir
+        self.json_file = json_file
+        self.name = name
+        self.coco = COCO(os.path.join(self.data_dir, "annotations", self.json_file))  # duplicate
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py]: Added exception if path does not exist.

File: src/super_gradients/training/utils/distributed_training_utils.py (hunk 11 above)
Comment (shaydeci) [reply src/super_gradients/training/utils/distributed_training_utils.py:130]: Used for waiting for rank 0 when caching images.

File: src/super_gradients/training/utils/distributed_training_utils.py (hunk 12 above)
Comment (shaydeci) [reply src/super_gradients/training/utils/distributed_training_utils.py]: removed

File: src/super_gradients/training/utils/distributed_training_utils.py (hunk 10 above)
Comment (shaydeci) [reply src/super_gradients/training/utils/distributed_training_utils.py:121]: Used for waiting for rank 0 when caching images.

File: src/super_gradients/training/sg_model/sg_model.py (hunk 9 above)
Comment (shaydeci) [reply src/super_gradients/training/sg_model/sg_model.py:935]: This is because for distributed samplers we had to call set_epoch each time we start a new epoch (so it would re-shuffle the data).
But sometimes we don't implement set epoch, or even use sampler (i.e use batch sampler), so we need to check before calling set_epoch.

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py (hunk 3 above)
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:152]: Not a problem. We have the dataset creation wrapped with wait_for_the_master() which waits for the caching of rank 0 to be completed. After that, the cache file already exists so this happens in rank 0 only.

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py (hunk 4 above)
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:132]: Not true, once the cache file is written (which is done once)- the different processes just sample from it according to the sampler.

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py (hunk 5 above)
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:213]: Derived from the image id as far as I know.

File: src/super_gradients/training/datasets/samplers/infinite_sampler.py (hunk 6 above)
Comment (shaydeci) [reply src/super_gradients/training/datasets/samplers/infinite_sampler.py]: removed and added license in https://github.com/Deci-AI/super-gradients/pull/223/commits/207848406e975b66dfeb5eaa801a0047ebefb670

File: src/super_gradients/training/datasets/samplers/infinite_sampler.py (hunk 7 above)
Comment (shaydeci) [reply src/super_gradients/training/datasets/samplers/infinite_sampler.py:1]: added license in https://github.com/Deci-AI/super-gradients/commit/207848406e975b66dfeb5eaa801a0047ebefb670

File: src/super_gradients/training/utils/detection_utils.py (hunk 8 above)
Comment (shaydeci) [reply src/super_gradients/training/utils/detection_utils.py]: removed

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py (hunk 2 above)
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py]: moved

File: src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py (hunk 1 above)
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_detection_yolox.py:87]: Yes, this happens sometimes.

**Review by ofrimasad — APPROVED** [review]
LGTM
